src/v/cluster/distributed_kv_stm.h

```diff
@@ -149,7 +149,8 @@ class distributed_kv_stm final : public raft::persisted_stm<> {
         _kvs = std::move(snap.kv_data);
     }
 
-    ss::future<raft::stm_snapshot> take_local_snapshot() override {
+    ss::future<raft::stm_snapshot>
+    take_local_snapshot(ssx::semaphore_units apply_units) override {
         auto holder = _gate.hold();
         auto units = co_await _snapshot_lock.hold_write_lock();
```

Comment:
I find it somewhat tricky to make sure `_snapshot_lock` and `_apply_lock` do not deadlock. I guess `replicate_and_wait` doesn't hold either of them on any of the remote nodes, but can it hold something that conflicts with them? Could you write some comment about it?